### PR TITLE
feat: [metrics] add optional metrics for --dsjson

### PR DESCRIPTION
**Pull request by lalo**

### CODE REVIEW

File: vowpalwabbit/metrics.cc [hunk 1]
```diff
@@ -33,7 +33,7 @@ struct metrics_data
   size_t predicted_not_first = 0;
 };
 
-void list_to_json_file(std::string filename, std::vector<std::tuple<std::string, size_t>>& metrics)
+void list_to_json_file(vw& all, std::string filename, std::vector<std::tuple<std::string, size_t>>& metrics)
```
Comment (olgavrou) [inline vowpalwabbit/metrics.cc:36]: is it worth passing a pointer to the `metrics` here instead of `all`?

**Review by olgavrou — APPROVED** [review]

File: vowpalwabbit/parse_example_json.h [hunk 2]
```diff
@@ -1726,11 +1732,28 @@ bool parse_line_json(vw* all, char* line, size_t num_chars, v_array<example*>& e
     VW::template read_line_decision_service_json<audit>(*all, examples, line, num_chars, false,
         reinterpret_cast<VW::example_factory_t>(&VW::get_unused_example), all, &interaction);
 
+    if (!interaction.eventId.empty())
+    {
+      if (all->example_parser->metrics->FirstEventId.empty())
```
Comment (ataymano) [inline vowpalwabbit/parse_example_json.h:1737]: looks like metrics is initialized conditionally (based on boolean flag), but there is no similar checks here. Or am I missing something?

Comment (lalo) [reply vowpalwabbit/parse_example_json.h:1737]: yes, right now it is initialized in the same place where the bool of line 1726 of this file gets set. I was expecting a test or some binding to break this contract of not calling into set_json_reader( but doesn't seem like it. Anyways, do you think I should move the initialization over here or just check if its null or not?

File: vowpalwabbit/parse_example_json.h [hunk 3]
```diff
@@ -1726,11 +1732,21 @@ bool parse_line_json(vw* all, char* line, size_t num_chars, v_array<example*>& e
     VW::template read_line_decision_service_json<audit>(*all, examples, line, num_chars, false,
         reinterpret_cast<VW::example_factory_t>(&VW::get_unused_example), all, &interaction);
 
+    if (!interaction.eventId.empty() && all->example_parser->metrics->FirstEventId.empty())
```
[the remark this answers is not in the document]
Comment (lalo) [reply vowpalwabbit/parse_example_json.h]: yes, in that case the contract would be to keep the first non-empty time and ID and to keep the last non-zero one for the last.

File: vowpalwabbit/metrics.cc (hunk 1 above)
Comment (lalo) [reply vowpalwabbit/metrics.cc]: done

File: vowpalwabbit/parse_example_json.h (hunk 2 above)
Comment (lalo) [reply vowpalwabbit/parse_example_json.h]: addressed this point, see latest diff

**Pull request merged**